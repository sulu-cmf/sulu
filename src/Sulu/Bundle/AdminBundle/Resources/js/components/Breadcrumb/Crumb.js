// @flow
import React from 'react';
import crumbStyles from './crumb.scss';

type Props = {
    onClick?: (value?: string | number) => void,
    children: string,
    value?: string | number,
};

export default class Crumb extends React.PureComponent<Props> {
    handleClick = () => {
        const {
            value,
            onClick,
        } = this.props;

        if (onClick) {
            onClick(value);
        }
    };

    render() {
        const {
            onClick,
            children,
        } = this.props;

        return (
            <button
                onClick={this.handleClick}
                disabled={!onClick}
                className={crumbStyles.crumb}
            >
                {children}
            </button>
        );
    }
}
