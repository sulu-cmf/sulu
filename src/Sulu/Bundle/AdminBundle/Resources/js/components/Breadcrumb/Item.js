// @flow
import React from 'react';
import itemStyles from './item.scss';

type Props = {
    onClick?: (value?: string | number) => void,
    children: string,
    value?: string | number,
};

export default class Item extends React.PureComponent<Props> {
    handleClick = () => {
        const {
            value,
            onClick,
        } = this.props;

        if (onClick) {
            onClick(value);
        }
    };

    render() {
        const {
            onClick,
            children,
        } = this.props;

        return (
            <button
                className={itemStyles.item}
                disabled={!onClick}
                onClick={this.handleClick}
            >
                {children}
            </button>
        );
    }
}
