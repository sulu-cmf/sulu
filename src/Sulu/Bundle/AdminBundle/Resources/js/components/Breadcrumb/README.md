The `Breadcrumb` indicates the current pageâ€™s location within a navigational hierarchy.

```
const handleClick = (value) => {
    alert(`You clicked on crumb ${value}`);
};

<Breadcrumb onItemClick={handleClick}>
    <Breadcrumb.Item>Root Directory</Breadcrumb.Item>
    <Breadcrumb.Item value={2}>...</Breadcrumb.Item>
    <Breadcrumb.Item>Current Directory</Breadcrumb.Item>
</Breadcrumb>
```
