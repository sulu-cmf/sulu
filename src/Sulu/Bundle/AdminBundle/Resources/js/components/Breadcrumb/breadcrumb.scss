@import '../../containers/Application/colors.scss';

$breadcrumbTextColor: $silver;

.breadcrumb {
    margin: 0 0 30px;
    padding: 0;
    list-style: none;
    display: flex;
    color: $breadcrumbTextColor;

    & > li {
        display: flex;
        align-items: center;
    }
}

.arrow {
    padding: 0 20px;
}
