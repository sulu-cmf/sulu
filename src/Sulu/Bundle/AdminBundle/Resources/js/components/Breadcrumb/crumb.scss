@import '../../containers/Application/colors.scss';

$crumbTextColor: $silver;
$crumbTextColorActive: $black;

.crumb {
    font-size: 30px;
    color: $crumbTextColor;
    background-color: transparent;
    border: none;
    cursor: pointer;
    padding: 0;

    &:hover,
    &:disabled {
        color: $crumbTextColorActive;
    }

    &:disabled {
        cursor: default;
    }
}
