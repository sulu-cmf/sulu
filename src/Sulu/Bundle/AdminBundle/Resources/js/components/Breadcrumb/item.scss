@import '../../containers/Application/colors.scss';

$itemTextColor: $silver;
$itemTextColorActive: $black;

.item {
    font-size: 30px;
    color: $itemTextColor;
    background-color: transparent;
    border: none;
    cursor: pointer;
    padding: 0;

    &:hover,
    &:disabled {
        color: $itemTextColorActive;
    }

    &:disabled {
        cursor: default;
    }
}
