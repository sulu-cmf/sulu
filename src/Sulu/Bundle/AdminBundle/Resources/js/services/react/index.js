// @flow
import buildHocDisplayName from './buildHocDisplayName';

export {buildHocDisplayName};
