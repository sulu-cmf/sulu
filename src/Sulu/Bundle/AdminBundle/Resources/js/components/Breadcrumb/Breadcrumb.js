// @flow
import React from 'react';
import type {ChildrenArray, Element} from 'react';
import Icon from '../Icon';
import Item from './Item';
import breadcrumbStyles from './breadcrumb.scss';

const ICON_ANGLE_RIGHT = 'su-angle-right';

type Props = {
    children: ChildrenArray<Element<typeof Item>>,
    onItemClick?: (value?: string | number) => void,
};

export default class Breadcrumb extends React.PureComponent<Props> {
    static Item = Item;

    createItems(originalItems: ChildrenArray<Element<typeof Item>>) {
        const childrenCount = React.Children.count(originalItems);

        return React.Children.map(originalItems, (item, index) => {
            const lastItem = (index === childrenCount - 1);

            return (
                <li>
                    {React.cloneElement(item, {
                        value: item.props.value,
                        onClick: (!lastItem) ? this.handleItemClick : undefined,
                    })}
                    {!lastItem &&
                        <Icon className={breadcrumbStyles.arrow} name={ICON_ANGLE_RIGHT} />
                    }
                </li>
            );
        });
    }

    handleItemClick = (value?: string | number) => {
        const {onItemClick} = this.props;

        if (onItemClick) {
            onItemClick(value);
        }
    };

    render() {
        const {
            children,
        } = this.props;
        const items = this.createItems(children);

        return (
            <ul className={breadcrumbStyles.breadcrumb}>
                {items}
            </ul>
        );
    }
}
