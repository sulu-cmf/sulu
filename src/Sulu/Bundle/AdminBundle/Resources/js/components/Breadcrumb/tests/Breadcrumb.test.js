// @flow
import {fireEvent, render, screen} from '@testing-library/react';
import React from 'react';
import Breadcrumb from '../Breadcrumb';

test('Render a Breadcrumb', () => {
    const clickSpy = jest.fn();
    const {container} = render(
        <Breadcrumb onItemClick={clickSpy}>
            <Breadcrumb.Item>
                Crumb 1
            </Breadcrumb.Item>
            <Breadcrumb.Item>
                Crumb 2
            </Breadcrumb.Item>
            <Breadcrumb.Item>
                Crumb 3
            </Breadcrumb.Item>
        </Breadcrumb>
    );
    expect(container).toMatchSnapshot();
});

test('Clicking on a clickable breadcrumb part should call a handler', () => {
    const clickSpy = jest.fn();
    const testValue = 2;
    const {debug} = render(
        <Breadcrumb onItemClick={clickSpy}>
            <Breadcrumb.Item>
                Crumb 1
            </Breadcrumb.Item>
            <Breadcrumb.Item value={testValue}>
                Crumb 2
            </Breadcrumb.Item>
            <Breadcrumb.Item>
                Crumb 3
            </Breadcrumb.Item>
        </Breadcrumb>
    );

    debug();

    const item = screen.queryByText('Crumb 2');
    fireEvent.click(item);

    expect(clickSpy).toHaveBeenCalledWith(testValue);
});
