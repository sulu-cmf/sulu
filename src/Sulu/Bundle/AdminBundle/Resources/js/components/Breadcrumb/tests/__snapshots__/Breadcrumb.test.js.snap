// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a Breadcrumb 1`] = `
<ul
  class="breadcrumb"
>
  <li>
    <button
      class="item"
    >
      Crumb 1
    </button>
    <span
      aria-label="chevron-right"
      class="arrow fa fa-chevron-right"
    />
  </li>
  <li>
    <button
      class="item"
    >
      Crumb 2
    </button>
    <span
      aria-label="chevron-right"
      class="arrow fa fa-chevron-right"
    />
  </li>
  <li>
    <button
      class="item"
      disabled=""
    >
      Crumb 3
    </button>
  </li>
</ul>
`;
